File: src/lib/gateway/registry.ts
import { GatewayClient } from './client'
import { type GatewayAdapter, resolveAdapter } from './adapter'
import { prisma } from '@/lib/db'
import { decrypt } from '@/lib/auth/encryption'
import type { ConfigGetResult, ConfigSchemaResult } from '@/types/gateway'

type ConnectionStatus = 'connecting' | 'connected' | 'disconnected' | 'error'

interface ManagedInstance {
  client: GatewayClient
  instanceId: string
  status: ConnectionStatus
}

const globalForRegistry = globalThis as unknown as {
  gatewayRegistry: GatewayRegistry
  registryInitialized?: boolean
}

export class GatewayRegistry {
  private instances = new Map<string, ManagedInstance>()

  async connect(instanceId: string, url: string, token: string): Promise<void> {
    // If already connected, disconnect first
    if (this.instances.has(instanceId)) {
      await this.disconnect(instanceId)
    }

    const client = new GatewayClient(url, token)
    const managed: ManagedInstance = { client, instanceId, status: 'connecting' }

    client.onStatusChange = (status) => {
      managed.status = status
    }

    client.onPermanentDisconnect = () => {
      managed.status = 'error'
      // Update DB status to ERROR (fire-and-forget)
      prisma.instance.update({
        where: { id: instanceId },
        data: { status: 'ERROR' },
      }).catch(console.error)
    }

    this.instances.set(instanceId, managed)
    await client.connect()
  }

  async disconnect(instanceId: string): Promise<void> {
    const managed = this.instances.get(instanceId)
    if (managed) {
      managed.client.disconnect()
      this.instances.delete(instanceId)
    }
  }

  getClient(instanceId: string): GatewayClient | undefined {
    return this.instances.get(instanceId)?.client
  }

  getAdapter(instanceId: string): GatewayAdapter | undefined {
    if (!this.instances.has(instanceId)) return undefined
    // Create fresh adapter to avoid stale class references from hot-reload
    return resolveAdapter()
  }

  getStatus(instanceId: string): ConnectionStatus | undefined {
    return this.instances.get(instanceId)?.status
  }

  getServerVersion(instanceId: string): string | null {
    return this.instances.get(instanceId)?.client.serverVersion ?? null
  }

  async request(instanceId: string, method: string, params?: Record<string, unknown>): Promise<unknown> {
    const client = this.getClient(instanceId)
    if (!client) {
      throw new Error(`Instance ${instanceId} is not connected`)
    }
    return client.request(method, params)
  }

  async getConfig(instanceId: string): Promise<ConfigGetResult> {
    const adapter = this.getAdapter(instanceId)
    const client = this.getClient(instanceId)
    if (!adapter || !client) {
      throw new Error(`Instance ${instanceId} is not connected`)
    }
    return adapter.getConfig(client)
  }

  async getSchema(instanceId: string): Promise<ConfigSchemaResult> {
    const adapter = this.getAdapter(instanceId)
    const client = this.getClient(instanceId)
    if (!adapter || !client) {
      throw new Error(`Instance ${instanceId} is not connected`)
    }
    return adapter.getSchema(client)
  }

  async patchConfig(
    instanceId: string,
    patch: Record<string, unknown>,
    baseHash: string,
  ): Promise<void> {
    const adapter = this.getAdapter(instanceId)
    const client = this.getClient(instanceId)
    if (!adapter || !client) {
      throw new Error(`Instance ${instanceId} is not connected`)
    }
    await adapter.patchConfig(client, patch, baseHash)
  }

  async checkHealth(instanceId: string): Promise<unknown> {
    const adapter = this.getAdapter(instanceId)
    const client = this.getClient(instanceId)
    if (!adapter || !client) {
      throw new Error(`Instance ${instanceId} is not connected`)
    }
    return adapter.getHealth(client)
  }

  isConnected(instanceId: string): boolean {
    return this.instances.get(instanceId)?.client.isConnected() ?? false
  }

  getConnectedIds(): string[] {
    return Array.from(this.instances.entries())
      .filter(([, m]) => m.status === 'connected')
      .map(([id]) => id)
  }

  async disconnectAll(): Promise<void> {
    const ids = Array.from(this.instances.keys())
    await Promise.all(ids.map(id => this.disconnect(id)))
  }
}

export const registry =
  globalForRegistry.gatewayRegistry ||
  (globalForRegistry.gatewayRegistry = new GatewayRegistry())

/**
 * Resolve the effective gateway URL for the current environment.
 * - Inside Docker (DOCKER_NETWORK set): use the DB URL as-is (container DNS)
 * - On host (no DOCKER_NETWORK): translate container DNS URLs to localhost + host port
 */
export function resolveGatewayUrl(inst: { gatewayUrl: string; dockerConfig: unknown }): string {
  const url = inst.gatewayUrl
  if (process.env.DOCKER_NETWORK) return url

  // On host: translate Docker-only hostnames and container DNS names
  try {
    const parsed = new URL(url.replace(/^ws/, 'http'))
    // host.docker.internal → 127.0.0.1 (same machine, just different DNS context)
    if (parsed.hostname === 'host.docker.internal') {
      return `ws://127.0.0.1:${parsed.port}`
    }
    // Container DNS name → localhost:hostPort (from dockerConfig)
    if (parsed.hostname !== '127.0.0.1' && parsed.hostname !== 'localhost') {
      const cfg = inst.dockerConfig as Record<string, unknown> | null
      if (cfg && typeof cfg.hostPort === 'number') {
        return `ws://127.0.0.1:${cfg.hostPort}`
      }
    }
  } catch {
    // invalid URL, return as-is
  }
  return url
}

// Lazy initialization: restore connections for all non-DISABLED instances.
// ERROR/OFFLINE instances are included because the container may have restarted
// since the status was set — skipping them would leave them stuck forever.
export async function ensureRegistryInitialized(): Promise<void> {
  if (globalForRegistry.registryInitialized) return
  globalForRegistry.registryInitialized = true

  try {
    const instances = await prisma.instance.findMany()

    await Promise.allSettled(
      instances.map(async (inst) => {
        try {
          const effectiveUrl = resolveGatewayUrl(inst)
          await registry.connect(inst.id, effectiveUrl, decrypt(inst.gatewayToken))
          // Connection succeeded — if instance was ERROR/OFFLINE, mark as DEGRADED
          // so the health check cycle can promote it to ONLINE on next success.
          if (inst.status === 'ERROR' || inst.status === 'OFFLINE') {
            await prisma.instance.update({
              where: { id: inst.id },
              data: { status: 'DEGRADED' },
            }).catch(console.error)
          }
        } catch (err) {
          console.error(`Failed to restore connection for instance ${inst.id}:`, err)
          // Only downgrade ONLINE/DEGRADED → ERROR; leave ERROR/OFFLINE as-is
          if (inst.status === 'ONLINE' || inst.status === 'DEGRADED') {
            await prisma.instance.update({
              where: { id: inst.id },
              data: { status: 'ERROR' },
            }).catch(console.error)
          }
        }
      })
    )
  } catch (err) {
    console.error('Failed to initialize gateway registry:', err)
  }

  // Start health checks + recovery in background (lazy import to avoid circular deps)
  import('./health').then(({ ensureHealthChecks }) =>
    ensureHealthChecks().catch(console.error),
  )
}

File: server/internal/service/gateway/registry.go
package gateway

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"go.uber.org/zap"
	"gorm.io/gorm"

	"github.com/szsip239/teamclaw/server/internal/model"
	"github.com/szsip239/teamclaw/server/internal/pkg/crypto"
)

// Registry manages persistent WebSocket connections to all OpenClaw Gateway instances.
// One Client is maintained per Instance; connections survive HTTP request lifetimes
// and are shared across all concurrent HTTP handlers.
type Registry struct {
	mu      sync.RWMutex
	clients map[string]*Client // instanceID → *Client
	status  map[string]ConnectionStatus

	db     *gorm.DB
	logger *zap.Logger
	enc    *crypto.Encryptor
}

// NewRegistry creates an empty registry. Call Initialize to connect instances.
func NewRegistry(db *gorm.DB, logger *zap.Logger, enc *crypto.Encryptor) *Registry {
	return &Registry{
		clients: make(map[string]*Client),
		status:  make(map[string]ConnectionStatus),
		db:      db,
		logger:  logger,
		enc:     enc,
	}
}

// Connect opens a gateway connection for the given instance.
// If a connection already exists it is disconnected first.
func (r *Registry) Connect(ctx context.Context, instanceID, url, token string) error {
	// Disconnect stale connection if any.
	r.mu.Lock()
	if existing, ok := r.clients[instanceID]; ok {
		existing.Disconnect()
	}
	r.mu.Unlock()

	client := NewClient(url, token, r.logger.With(zap.String("instanceId", instanceID)))

	client.OnStatusChange = func(status ConnectionStatus) {
		r.mu.Lock()
		r.status[instanceID] = status
		r.mu.Unlock()
	}

	client.OnPermanentDisconnect = func() {
		// Mark instance as ERROR in DB (fire-and-forget).
		r.db.Model(&model.Instance{}).
			Where("id = ?", instanceID).
			Update("status", model.InstanceStatusError)
	}

	r.mu.Lock()
	r.clients[instanceID] = client
	r.status[instanceID] = StatusConnecting
	r.mu.Unlock()

	if err := client.Connect(ctx); err != nil {
		r.mu.Lock()
		delete(r.clients, instanceID)
		delete(r.status, instanceID)
		r.mu.Unlock()
		return fmt.Errorf("registry: connect %s: %w", instanceID, err)
	}
	return nil
}

// Disconnect closes the connection for the given instance.
func (r *Registry) Disconnect(instanceID string) {
	r.mu.Lock()
	client := r.clients[instanceID]
	delete(r.clients, instanceID)
	delete(r.status, instanceID)
	r.mu.Unlock()

	if client != nil {
		client.Disconnect()
	}
}

// GetClient returns the Client for the given instance, or nil if not connected.
func (r *Registry) GetClient(instanceID string) *Client {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.clients[instanceID]
}

// IsConnected returns true if the instance has a live, authenticated connection.
func (r *Registry) IsConnected(instanceID string) bool {
	r.mu.RLock()
	client := r.clients[instanceID]
	r.mu.RUnlock()
	if client == nil {
		return false
	}
	return client.IsConnected()
}

// GetStatus returns the current connection status for the instance.
func (r *Registry) GetStatus(instanceID string) (ConnectionStatus, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	s, ok := r.status[instanceID]
	return s, ok
}

// GetServerVersion returns the gateway version string from the hello-ok handshake.
func (r *Registry) GetServerVersion(instanceID string) string {
	r.mu.RLock()
	client := r.clients[instanceID]
	r.mu.RUnlock()
	if client == nil {
		return ""
	}
	return client.ServerVersion()
}

// GetConnectedIDs returns all instance IDs that currently have an authenticated connection.
func (r *Registry) GetConnectedIDs() []string {
	r.mu.RLock()
	defer r.mu.RUnlock()
	ids := make([]string, 0)
	for id, client := range r.clients {
		if client.IsConnected() {
			ids = append(ids, id)
		}
	}
	return ids
}

// Request sends a method call to the gateway for the given instance.
func (r *Registry) Request(ctx context.Context, instanceID, method string, params any) (json.RawMessage, error) {
	client := r.GetClient(instanceID)
	if client == nil {
		return nil, fmt.Errorf("registry: instance %s is not connected", instanceID)
	}
	return client.Request(ctx, method, params, 0)
}

// DisconnectAll gracefully closes all open connections.
func (r *Registry) DisconnectAll() {
	r.mu.Lock()
	clients := make(map[string]*Client, len(r.clients))
	for id, c := range r.clients {
		clients[id] = c
	}
	r.clients = make(map[string]*Client)
	r.status = make(map[string]ConnectionStatus)
	r.mu.Unlock()

	for _, c := range clients {
		c.Disconnect()
	}
}

// Initialize loads all instances from the database and connects to each one.
// Instances with ERROR or OFFLINE status are also attempted — the health checker
// will promote them to ONLINE on success or keep them ERROR on failure.
// Connection errors are logged but do not abort initialization.
func (r *Registry) Initialize(ctx context.Context) {
	var instances []model.Instance
	if err := r.db.Find(&instances).Error; err != nil {
		r.logger.Error("registry: failed to load instances", zap.Error(err))
		return
	}

	var wg sync.WaitGroup
	sem := make(chan struct{}, 5) // max 5 concurrent connects

	for _, inst := range instances {
		inst := inst
		wg.Add(1)
		go func() {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			token, err := r.enc.Decrypt(inst.GatewayToken)
			if err != nil {
				r.logger.Error("registry: failed to decrypt token",
					zap.String("instanceId", inst.ID), zap.Error(err))
				return
			}

			connCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
			defer cancel()

			if err := r.Connect(connCtx, inst.ID, inst.GatewayURL, token); err != nil {
				r.logger.Warn("registry: initial connect failed",
					zap.String("instanceId", inst.ID),
					zap.String("url", inst.GatewayURL),
					zap.Error(err))
				// Downgrade ONLINE/DEGRADED → ERROR; leave ERROR/OFFLINE as-is.
				if inst.Status == model.InstanceStatusOnline || inst.Status == model.InstanceStatusDegraded {
					r.db.Model(&inst).Update("status", model.InstanceStatusError)
				}
				return
			}

			// Connection succeeded — if instance was ERROR/OFFLINE, mark DEGRADED
			// so the health checker can promote it to ONLINE on first success.
			if inst.Status == model.InstanceStatusError || inst.Status == model.InstanceStatusOffline {
				r.db.Model(&inst).Update("status", model.InstanceStatusDegraded)
			}
		}()
	}

	wg.Wait()
	r.logger.Info("registry: initialization complete",
		zap.Int("total", len(instances)),
		zap.Int("connected", len(r.GetConnectedIDs())),
	)
}
